### PR TITLE
LibraryActions work — i.e., public library API should work!

### PR DESCRIPTION
**Pull request by andrewconner**
Bit more to do, but this is the bulk of the backend work to make public libraries work. @ahsu1230 @yipingliao 

@raymondng Actually want to bounce an idea off you about the new Request API

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 1]
```diff
@@ -44,79 +45,56 @@ class LibraryController @Inject() (
       case Left(LibraryFail(message)) =>
         Future.successful(BadRequest(Json.obj("error" -> message)))
       case Right(newLibrary) =>
-        libraryCommander.createFullLibraryInfo(request.userId, newLibrary).map { lib =>
+        libraryCommander.createFullLibraryInfo(Some(request.userId), newLibrary).map { lib =>
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:48]: small thing, but could use request.userIdOpt?

**Comment by ahsu1230** [comment]
really liking the UserAction andThen LibraryActions - makes the controller API methods even thinner! :+1: good job!

**Comment by raymondng** [comment]
Sure just grab me whenever -- would love to hear it.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 2]
```diff
@@ -50,7 +50,7 @@ class LibraryCommander @Inject() (
     implicit val publicIdConfig: PublicIdConfiguration,
     clock: Clock) extends Logging {
 
-  def createFullLibraryInfo(userId: Id[User], library: Library): Future[FullLibraryInfo] = {
+  def createFullLibraryInfo(userIdOpt: Option[Id[User]], library: Library): Future[FullLibraryInfo] = {
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala]: might help readers if we include the user’s role in the name of this argument, like `viewerIdOpt` or `viewerUserIdOpt`

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:48]: No need. Other uses are using an implicit conversion to pattern match on the request to extract the `userId` if it's a `UserRequest`. In this case, it's always there. imho, better not to confuse future readers about a possibility of `None` here.

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/LibraryAccessActions.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  implicit private val implicitPublicId = publicIdConfig
+  val libraryCommander: LibraryCommander
+
+  def LibraryViewAction(id: PublicId[Library]) = new ActionFilter[MaybeUserRequest] {
+    def filter[A](input: MaybeUserRequest[A]): Future[Option[Result]] = Future.successful(lookupViewAccess(id, input))
+  }
+
+  def LibraryWriteAction(id: PublicId[Library]): ActionFilter[UserRequest] = new ActionFilter[UserRequest] {
+    def filter[A](input: UserRequest[A]): Future[Option[Result]] = Future.successful(lookupWriteAccess(id, input))
+  }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/LibraryAccessActions.scala:30]: Cool. I guess these are package-specific? I may put a copy of these in `controllers.ext` without the passphrase stuff.

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/LibraryAccessActions.scala:30]: They can be used anywhere on shoebox. The auth token / passphrase stuff shouldn't cause problems, but may introduce a very slight performance hit. Your call.

**Pull request merged**